athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchMetadataHandler.java

```diff
@@ -0,0 +1,282 @@
+ */
+public class ElasticsearchMetadataHandler
+        extends GlueMetadataHandler
+{
+    private static final Logger logger = LoggerFactory.getLogger(ElasticsearchMetadataHandler.class);
+
+    // Used to denote the 'type' of this connector for diagnostic purposes.
+    private static final String SOURCE_TYPE = "elasticsearch";
+
+    //The Env variable name used to indicate that we want to disable the use of Glue DataCatalog for supplemental
+    //metadata and instead rely solely on the connector's schema inference capabilities.
+    private static final String GLUE_ENV = "disable_glue";
+
+    private final AWSGlue awsGlue;
+    private final AwsRestHighLevelClientFactory clientFactory;
+
+    public ElasticsearchMetadataHandler()
+    {
+        //Disable Glue if the env var is present and not explicitly set to "false"
+        super((System.getenv(GLUE_ENV) != null && !"false".equalsIgnoreCase(System.getenv(GLUE_ENV))), SOURCE_TYPE);
+        this.awsGlue = getAwsGlue();
+        this.clientFactory = ElasticsearchHelper.getClientFactory();
+
+        ElasticsearchHelper.setDomainMapping();
+    }
+
+    @VisibleForTesting
+    protected ElasticsearchMetadataHandler(AWSGlue awsGlue,
+                                           EncryptionKeyFactory keyFactory,
+                                           AWSSecretsManager awsSecretsManager,
+                                           AmazonAthena athena,
+                                           String spillBucket,
+                                           String spillPrefix,
+                                           AwsRestHighLevelClientFactory clientFactory)
+    {
+        super(awsGlue, keyFactory, awsSecretsManager, athena, SOURCE_TYPE, spillBucket, spillPrefix);
+        this.awsGlue = awsGlue;
+        this.clientFactory = clientFactory;
+
+        ElasticsearchHelper.setDomainMapping();
+    }
+
+    /**
+     * Used to get the list of domains (aka databases) for the Elasticsearch service.
+     * @param allocator Tool for creating and managing Apache Arrow Blocks.
+     * @param request Provides details on who made the request and which Athena catalog they are querying.
+     * @return A ListSchemasResponse which primarily contains a Set<String> of schema names and a catalog name
+     * corresponding the Athena catalog that was queried.
+     */
+    @Override
+    public ListSchemasResponse doListSchemaNames(BlockAllocator allocator, ListSchemasRequest request)
+    {
+        logger.info("doListSchemaNames: enter - " + request);
+
+        if (ElasticsearchHelper.isAutoDiscoverEndpoint()) {
+            // Since adding/deleting domains in Amazon Elasticsearch Service doesn't re-initialize
+            // the connector, the domainMap needs to be refreshed each time for Amazon ES.
+            ElasticsearchHelper.setDomainMapping();
+        }
+
+        return new ListSchemasResponse(request.getCatalogName(), ElasticsearchHelper.getDomainList());
```

Comment:
Why is getDomainList static vs having an instance of some Elasticsearch object here that you call? in general, statics should be used exceedingly sparingly for things that are stateful.

Reply:
Domain mapping is static for the reasons stated in the previous comment. Static implementation getDomainList() is simply a way of retrieving the key-set from that mapping. Of, course if it becomes part of the Record and Metadata handlers' classes, it would mean duplication.

Reply:
Made ElasticsearchHelper a singleton rather than a static class.